mesg-foundation/engine
# Workflow graph structure

Update the structure of the workflow with a graph approach.

The workflow definition is now created based on a list of nodes and a list of edges that connects these nodes. 
Every node need to be unique in the workflow and will be identifiable. With the node identifiable we can now store the current step when creating execution and can drop the height support that was not valid when a workflow was triggered by a result of the same workflow and the same task.

With this new `stepID` we can simply find the children of this node in the workflow definition graph and execute all the children's tasks.

- SDK of execution updated to accept the `stepID`
- API updated with the new workflow struct and execution stepID

## TO TEST

Create and start the services with the cli or API.

```json
// ServiceA
{"name":"serviceA","tasks":[{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskX"},{"key":"emits"}],"events":[{"data":[{"type":"String","key":"message"}],"key":"eventX"}],"configuration":{},"hash":"3kWX7CYtNwbeb1h35HiNAPctnh94MU8zstT7PSuusxyy","sid":"serviceA","source":"QmUojyMjo2JUvhktLU4royUBDEsVzU4z1onPovGpVQAFDW"}

// ServiceB
{"name":"serviceB","tasks":[{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskY"}],"configuration":{},"hash":"8ExNFweHXavpZSE9iTwjBFSDWzKcRpCXEc4BuX1UwFcg","sid":"serviceB","source":"Qmf2e92jNxxmYyjxEFrCBJdCNUS1LKNWet2ZQt3ByEUpAP"}

// ServiceC
{"name":"serviceC","tasks":[{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskZ"},{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskA"}],"configuration":{},"hash":"qTtisJG3Bso5rRxSHhnrQSkJpC3JGePD3ovvuEGVgeg","sid":"serviceC","source":"QmPuHX3LumnDkrxY3wUR8J2J9xmXGw8yTVy5jf7P24MprN"}
```

Create one or all the following workflow(s) with the API 

eg: `grpcurl -plaintext -d "$(cat result.json| jq -c .)" localhost:50052 api.Workflow/Create`

```json
// result.json
{"key":"result","trigger":{"type":2,"instanceHash":"A7712VMAiT2ehXV1nHHUnkvyTbZKVHeFxBZZj4Gx7MZk","key":"taskX","initilNode":"task1"},"nodes":[{"key":"task1","instanceHash":"DUto1J3ZAheVL6PGrwYTtGF4Bnw15CQocVdEmqkHSGbB","taskKey":"taskY"}, {"key":"task2","instanceHash":"FjFzXXme2dkLBsgPk2xgyiPD6dX5EN1dBLGqW6G9tStf","taskKey":"taskZ"}],"edges":[{"src":"task1","dst":"task2"}]}

// event.json
{"key":"event","trigger":{"type":1,"instanceHash":"A7712VMAiT2ehXV1nHHUnkvyTbZKVHeFxBZZj4Gx7MZk","key":"eventX","initialNode":"task1"},"nodes":[{"key":"task1","instanceHash":"DUto1J3ZAheVL6PGrwYTtGF4Bnw15CQocVdEmqkHSGbB","taskKey":"taskY"},{"key":"task2","instanceHash":"FjFzXXme2dkLBsgPk2xgyiPD6dX5EN1dBLGqW6G9tStf","taskKey":"taskZ"}],"edges":[{"src":"task1","dst":"task2"}]}

// result-from-other-workflow
{"key":"result-from-other-workflow","trigger":{"type":2,"instanceHash":"DUto1J3ZAheVL6PGrwYTtGF4Bnw15CQocVdEmqkHSGbB","key":"taskY","initialNode":"task1"},"nodes":[{"key":"task1","instanceHash":"FjFzXXme2dkLBsgPk2xgyiPD6dX5EN1dBLGqW6G9tStf","taskKey":"taskZ"}],"edges":[]}
```

Then you see the workflow by running either an event with 
```bash
mesg-cli service:execute serviceA emits
```
or a task
```bash
mesg-cli service:execute serviceA taskX --data message=hello
```
